### ISSUE TITLE
run karmada join failed

### ISSUE THREAD

**Issue by zirain** [karmada-io/karmada#1836]
**What happened**:

run `karmada-ctl join` after `karmada init` faild.

**What you expected to happen**:

cluster join karmada as a member.

**How to reproduce it (as minimally and precisely as possible)**:

run `karmada-ctl join` once `karmada init` completed.
```console
Step 4: Show members of karmada                                                                                                                                              
(In karmada)~# kubectl  --kubeconfig /etc/karmada/karmada-apiserver.config get clusters
# karmada init completed here, run karmada join immediately

$ karmada-ctl join member1 --kubeconfig=/etc/karmada/karmada-apiserver.config     --cluster-kubeconfig=/root/.kube/members.config     --cluster-context=member1
W0519 17:06:55.846781  457653 cluster.go:96] failed to retrieve cluster(). error: the server is currently unable to handle the request (get clusters.cluster.karmada.io member1)
Error: failed to create cluster(member1) object. error: the server is currently unable to handle the request (get clusters.cluster.karmada.io member1)
ERRO[2022-05-19 17:06:55] failed to join cluster: exit status 1        
ERRO[2022-05-19 17:06:55] join execute error: exit status 1            
```

there is residual secrets in karmada-cluster namespace, and will reject re-join cluster.
```
NAMESPACE         NAME                                    TYPE                                  DATA   AGE
karmada-cluster   member1                                 Opaque                                2      7m30s
karmada-cluster   member1-impersonator                    Opaque                                1      7m30s
karmada-cluster   member2                                 Opaque                                2      7m28s
karmada-cluster   member2-impersonator                    Opaque                                1      7m28s
kube-system       generic-garbage-collector-token-88qgn   kubernetes.io/service-account-token   3      7m40s
kube-system       namespace-controller-token-6zbxd        kubernetes.io/service-account-token   3      7m40s
```

the failure message is not friendly for user:
```console
W0519 17:15:56.009420  458439 cluster.go:106] failed to create cluster(member1). error: Cluster.cluster.karmada.io "member1" is invalid: [spec.secretRef.namespace: Required value, spec.secretRef.name: Required value, spec.impersonatorSecretRef.namespace: Required value, spec.impersonatorSecretRef.name: Required value]
W0519 17:15:56.009469  458439 cluster.go:50] failed to create cluster(member1). error: Cluster.cluster.karmada.io "member1" is invalid: [spec.secretRef.namespace: Required value, spec.secretRef.name: Required value, spec.impersonatorSecretRef.namespace: Required value, spec.impersonatorSecretRef.name: Required value]
Error: failed to create cluster(member1) object. error: Cluster.cluster.karmada.io "member1" is invalid: [spec.secretRef.namespace: Required value, spec.secretRef.name: Required value, spec.impersonatorSecretRef.namespace: Required value, spec.impersonatorSecretRef.name: Required value]
ERRO[2022-05-19 17:15:56] failed to join cluster: exit status 1        
ERRO[2022-05-19 17:15:56] join execute error: exit status 1    
```

**Anything else we need to know?**:

**Environment**:
- Karmada version: v1.1.1
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`): v1.1.1
- Others:

**Comment by hanweisen** [comment]
`cluster` is handle by `karmada-aggregated-apiserver.` May be this is cause by your `karmada-aggregated-apiserver` is not ready. I reproduce your problem by set 'karmada-aggregated-apiserver' replice zero
```
[root@harbor ~]#  kubectl get deployment -n karmada-system
NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
karmada-aggregated-apiserver          0/0     0            0           32d
karmada-apiserver                     1/1     1            1           32d
karmada-controller-manager            1/1     1            1           32d
karmada-scheduler                     1/1     1            1           32d
karmada-scheduler-estimator-member1   2/2     2            2           25d
karmada-scheduler-estimator-member2   2/2     2            2           25d
karmada-webhook                       1/1     1            1           32d
kube-controller-manager               1/1     1            1           32d
[root@harbor ~]#  kubectl get deployment -n karmada-system
NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
karmada-aggregated-apiserver          0/0     0            0           32d
karmada-apiserver                     1/1     1            1           32d
karmada-controller-manager            1/1     1            1           32d
karmada-scheduler                     1/1     1            1           32d
karmada-scheduler-estimator-member1   2/2     2            2           25d
karmada-scheduler-estimator-member2   2/2     2            2           25d
karmada-webhook                       1/1     1            1           32d
kube-controller-manager               1/1     1            1           32d
[root@harbor ~]#  kubectl karmada join single --kubeconfig=/etc/karmada/karmada-apiserver.config     --cluster-kubeconfig=/root/.kube/config     --cluster-context=single
W0519 21:23:10.288286   31861 cluster.go:96] failed to retrieve cluster(). error: the server is currently unable to handle the request (get clusters.cluster.karmada.io single)
Error: failed to create cluster(single) object. error: the server is currently unable to handle the request (get clusters.cluster.karmada.io single)
```

**Comment by zirain** [comment]
> I start karmada-aggregated-apiserver cluster can be join successfully

maybe `karmada init` should exit unless `aa` is ready?

**Comment by XiShanYongYe-Chang** [comment]
> maybe `karmada init` should exit unless `aa` is ready?

It is not enough to judge that the aa is ready. We should also judge that the cluster api is ready:
```
# kubectl api-resources | grep clusters
clusters                                                cluster.karmada.io/v1alpha1            false        Cluster
```

**Comment by zirain** [comment]
> > maybe `karmada init` should exit unless `aa` is ready?
> 
> It is not enough to judge that the aa is ready. We should also judge that the cluster api is ready:
> 
> ```
> # kubectl api-resources | grep clusters
> clusters                                                cluster.karmada.io/v1alpha1            false        Cluster
> ```

as we talk offline, we can change `aa` ready after cluster api is registered, and `karmdactl init` should wait for `aa` ready?

**Comment by anu491** [comment]
> > the failure message is not friendly for user:
> 
> ```
> diff --git a/pkg/util/cluster.go b/pkg/util/cluster.go
> index 4d2b7cda..8f26a7cb 100644
> --- a/pkg/util/cluster.go
> +++ b/pkg/util/cluster.go
> @@ -103,7 +103,7 @@ func GetClusterWithKarmadaClient(client karmadaclientset.Interface, name string)
>  func createCluster(controlPlaneClient karmadaclientset.Interface, cluster *clusterv1alpha1.Cluster) (*clusterv1alpha1.Cluster, error) {
>         newCluster, err := controlPlaneClient.ClusterV1alpha1().Clusters().Create(context.TODO(), cluster, metav1.CreateOptions{})
>         if err != nil {
> -               klog.Warningf("failed to create cluster(%s). error: %v", cluster.Name, err)
> +               klog.Warningf("failed to create cluster(%s). error: %v. May be karmada-aggregated-apiserver is not ready.", cluster.Name, err)
>                 return nil, err
>         }
> ```

I agree this change should be made . The initial message doesn't indicate any solution or statement which the user should be provided with , so that the user can take any necessary steps. The suggested change looks pretty good here and I'll make the change soon ! Thanks @hanweisen @zirain for pointing this out !

**Comment by anu491** [comment]
> The suggested change looks pretty good here and I'll make the change soon ! Thanks @hanweisen @zirain for pointing this out
 
A small doubt here I had here is that whether the same change should be made on line number 50 (just because the function calls are related I thought that we could also maybe consider the following change)

```
diff --git a/pkg/util/cluster.go b/pkg/util/cluster.go
index 4c137eac..e9d1905d 100644
--- a/pkg/util/cluster.go
+++ b/pkg/util/cluster.go
@@ -47,7 +47,7 @@ func CreateClusterObject(controlPlaneClient *karmadaclientset.Clientset, cluster
        }
 
        if cluster, err = createCluster(controlPlaneClient, clusterObj); err != nil {
-               klog.Warningf("failed to create cluster(%s). error: %v", clusterObj.Name, err)
+               klog.Warningf("failed to create cluster(%s). error: %v. May be karmada-aggregated-apiserver is not ready.", clusterObj.Name, err)
                return nil, err
        }
```

**Comment by XiShanYongYe-Chang** [comment]
> I can not imagine other situation will cause this error.I think we can make a friendly error message by modify util/cluster.go

This error is caused by the failure to join the cluster for the first time, but residual resources are still on the control plane, not because AA is not ready.

**Comment by anu491** [comment]
> This error is caused by the failure to join the cluster for the first time, but residual resources are still on the control plane, not because AA is not ready.
> 
> > klog.Warningf("failed to create cluster(%s). error: %v. May be karmada-aggregated-apiserver is not ready.", clusterObj.Name, err)
> 
> so the addition of this description is not accurate.

Okay , I went through the issue again and tried to understand what you're trying to convey. I realize the addition might not be appropriate . What would you suggest for an accurate and more user friendly addition here ?

**Comment by zirain** [comment]
I think there's nothing to do here, the root cause is here:
https://github.com/karmada-io/karmada/blob/8ea4c4d8819dc4c4b5fa885e3b24bb2fef910dbb/pkg/util/secret.go#L46

seems this's fixed in 
https://github.com/karmada-io/karmada/blob/c8e1b78ea1865241fafeacb7727681d53eb46741/pkg/util/secret.go#L46

@XiShanYongYe-Chang

**Comment by XiShanYongYe-Chang** [comment]
@zirain thanks~ It has been fixed by #1306

/close
